kubernetes/cache/cache.go

```diff
@@ -123,10 +131,35 @@ func NewKialiCache() (KialiCache, error) {
 	}
 	log.Tracef("[Kiali Cache] cacheIstioTypes %v", cacheIstioTypes)
 
-	stopChan := make(map[string]chan struct{})
+	// To determine if the cache is cluster-scoped or namespace-scoped,
+	// we check if Kiali has cluster-scoped permissions for each of the
+	// cache types. If Kiali does not have cluster-scoped permissions
+	// for any of the cache types, it is considered namespace-scoped.
+	// TODO: Can probably optimize this by paralellizing the cani checks.
```

Comment:
I'm inclined to assume this comes from the configuration and avoid determining it dynamically.
Then, if some situation fails (i.e. no right permissions, or for some reason, some API is not present), I think it's better to have a clean error message about an expected situation than having this.

I guess it will get fewer moving parts that, in the long term, may help to maintain better.

But the logic is correct, I think it's used in other places as well.

Reply:
> Then, if some situation fails (i.e. no right permissions, or for some reason, some API is not present), I think it's better to have a clean error message about an expected situation than having this.

That's a good point. Someone may want the cluster scoped cache but not realize that they dropped to a namespace scoped cache because of lack of cluster permissions for a particular type. A large motivation for determining it dynamically was to avoid adding more cache related config options. My assumption was that users will just want the caching option that is faster and more efficient as long as they have the correct permissions for it. Kiali can choose this for them without needing an extra config option but to your point they may want the cluster scoped cache but be unaware that they are running with the namespace scoped cache if kiali transparently drops to the namespaced cache. I think that's more of a problem of informing the user what is happening within Kiali though.

Would a user ever _choose_ to have a namespace scoped cache when they could have a cluster scoped cache that is more efficient? I personally don't think so and Kiali could just the better option when it's available. Adding more config options adds complexity too it just pushes it more onto the user and into the deployment/operator/docs.

We could include the cache scope in the kiali status to make it more visible to users what was chosen? (namespace vs. cluster scoped)

Reply:
I doubt users would care about scoping (just repsonse time), but if we think it would be useful for debugging then yes.

Reply:
I think that the operator should have the responsibility to populate that value.

Dynamic configuration as this one typically presents more edges that create issues in the long term; as an example, Kiali also had some "auto-discovery" logic for the "tracing" service, and that was a source of a lot of combinations that users presented.

If you are comfortable with the discovery logic, to me, it's ok; just sharing my thoughts here.

Reply:
> I think that the operator should have the responsibility to populate that value.
>
> Dynamic configuration as this one typically presents more edges that create issues in the long term; as an example, Kiali also had some "auto-discovery" logic for the "tracing" service, and that was a source of a lot of combinations that users presented.

That's a fair point. It could very well add some edge cases in the future. Having a config option that defaults to `clusterScope: enabled` would accomplish much the same thing in terms of ease of use. I'll remove the autodiscovery code in favor of a config option defaulting the cache to cluster scoped.